### PR TITLE
Uvs 90: Create fileStatusCoordinator worker in the uvs repo

### PR DESCRIPTION
**Pull request by VolodymyrFihurniak**

### CODE REVIEW

File: .github/workflows/CI-debug.yml [hunk 1]
```diff
@@ -29,3 +29,16 @@ jobs:
...
+    steps:
+      - name: Clone repository
+        uses: actions/checkout@v4
+      - name: Update docker compose
+        run: sudo curl -L "https://github.com/docker/compose/releases/download/v2.26.1/docker-compose-linux-x86_64" -o /usr/local/bin/docker-compose
+      - name: Change permissions
+        run: sudo chmod +x /usr/local/bin/docker-compose
+      - name: Print docker compose version
+        run: docker-compose --version
+      - name: Run file coordinator
```
Comment (denbykov) [inline .github/workflows/CI-debug.yml:43]: You are not running this application, just building it

File: docker-compose.yml [hunk 2]
```diff
@@ -64,25 +64,40 @@ services:
...
+    networks:
+      - up-vibe-network
+  file-coordinator:
+    build:
+      context: ./services/file-coordinator
+      dockerfile: ./Dockerfile
+    command: -c "npm run start"
+    container_name: file-coordinator
+    volumes:
+      - ./services/file-coordinator/configs:/opt/upVibe/services/file-coordinator/configs
```
Comment (denbykov) [inline docker-compose.yml:99]: Word config should be in singular form

File: services/file-coordinator/src/business/fileWorker.ts
Comment (denbykov) [inline services/file-coordinator/src/business/fileWorker.ts:1]: You do not need this class

File: services/file-coordinator/src/business/tagMappingWorker.ts
Comment (denbykov) [inline services/file-coordinator/src/business/tagMappingWorker.ts:1]: You do not need this class

File: services/file-coordinator/src/business/tagWorker.ts
Comment (denbykov) [inline services/file-coordinator/src/business/tagWorker.ts:1]: You do not need this class

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+class FileCoordinatorWorker {
+  private db: FileCoordinatorDatabase;
+  private logger: Logger;
+  constructor(db: FileCoordinatorDatabase, logger: Logger) {
+    this.db = db;
+    this.logger = logger;
+  }
+
+  public verifyFileStatus = (file: FileDTO): boolean => {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:18]: This function is redundant

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      return true;
+    }
+    return false;
+  };
+
+  public findFileAndCheckStatus = async (id: string): Promise<boolean> => {
+    const file = await this.db.getFileById(id);
+    const isDownloaded = this.verifyFileStatus(file!);
+    if (isDownloaded) {
+      this.logger.info(`File ${id} is downloaded`);
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:29]: This logs can only be useful in the debug or trace mode

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  public verifyFileStatus = (file: FileDTO): boolean => {
+    if (file.status === Status.Downloaded) {
+      return true;
+    }
+    return false;
+  };
+
+  public findFileAndCheckStatus = async (id: string): Promise<boolean> => {
+    const file = await this.db.getFileById(id);
+    const isDownloaded = this.verifyFileStatus(file!);
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:27]: Replace with const isDownloaded = (file.status === Status.Downloaded)

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:1]: All logs in this file are unsuitable for information level, as they will flood logs with useless spam. Change the level to debug or trance, or just remove them

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    this.logger.info(`Updating tag mapping by id: ${tagMapping.id}`);
+    await this.db.updateTagMappingById(tagMapping);
+  };
+
+  public validateTagStatus = (tag: TagDTO): boolean => {
+    this.logger.info(`Validating tag: ${tag.id}`);
+    return tag.status === Status.Completed || tag.status === Status.Error;
+  };
+
+  public getTagByFileId = async (fileId: string): Promise<TagDTO[]> => {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:113]: Rename to getTagsByFileId

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+  public processFile = async (
+    fileId: string,
+    userId: string,
+    deviceId: UUID
+  ): Promise<void> => {
+    const isDownloaded = await this.findFileAndCheckStatus(fileId);
+    if (isDownloaded) {
+      const tags = await this.getTagByFileId(fileId);
+      const validTags = tags.filter((tag) => {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:126]: Why are you filtering tags?

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        return this.validateTagStatus(tag);
+      });
+
+      if (validTags.length === 0) {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:130]: Does not comply with AC 2.1 at all

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+  };
+
+  public getTagByFileId = async (fileId: string): Promise<TagDTO[]> => {
+    this.logger.info(`Getting tags by file id: ${fileId}`);
+    return this.db.getTagByFileId(fileId);
+  };
+
+  public processFile = async (
+    fileId: string,
+    userId: string,
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:120]: Why do you have userIf and deviceId in parameters?

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+  public processFile = async (fileId: string): Promise<void> => {
+    const isDownloaded = await this.findFileAndCheckStatus(fileId);
+
+    if (!isDownloaded) {
+      return;
+    }
+    const tags = await this.db.getTagsByFileId(fileId);
+
+    if (!tags.every(this.validateTagStatus)) {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:65]: Consider using lambdas in such cases

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    return tag.status === Status.Completed || tag.status === Status.Error;
+  };
+
+  public processFile = async (fileId: string): Promise<void> => {
+    const isDownloaded = await this.findFileAndCheckStatus(fileId);
+
+    if (!isDownloaded) {
+      return;
+    }
+    const tags = await this.db.getTagsByFileId(fileId);
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:63]: You are lacking tags.size() > 1 check

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    if (!tags.every(this.validateTagStatus)) {
+      return;
+    }
+
+    const tagMapping = await this.db.getTagMappingByFileId(fileId);
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:69]: There is more than one mapping that you should read here

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+    if (tagMapping.fixed) {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:71]: The rest is obviously wrong

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  };
+
+  public mapTagMappingByMappingPriority = (
+    tagMappingPriority: TagMappingPriorityDTO,
+    tagMapping: TagMappingDTO,
+    tags: TagDTO[]
+  ): TagMappingDTO => {
+    const findFirstTag = (ids: string[]): string => {
+      const tag = ids.find((id) => {
+        return this.findTagById(tags, id) !== undefined;
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:32]: Use lambda here

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+      });
+      return tag || '';
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:34]: Why are you returning an empty string?

File: services/file-coordinator/src/business/fileCoordinatorWorker.ts [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  private findTagById = (tags: TagDTO[], id: string): TagDTO | undefined => {
+    return tags.find((tag) => tag.id === id);
+  };
+
+  public mapTagMappingByMappingPriority = (
+    tagMappingPriority: TagMappingPriorityDTO,
+    tagMapping: TagMappingDTO,
+    tags: TagDTO[]
+  ): TagMappingDTO => {
+    const findFirstTag = (ids: string[]): string => {
```
Comment (denbykov) [inline services/file-coordinator/src/business/fileCoordinatorWorker.ts:30]: This function does not make any sense to me

**Review by denbykov — DISMISSED** [review]

File: plugins/FileCoordinatorPlugin/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    this.dataLogger = dataLogger;
+    this.dataLogger.info(
+      `[${this.pluginName}] Plugin ${this.pluginName} loaded`
+    );
+  }
+  public coordinateFile =  (fileId: string): void => {
+    this.dataLogger.info(`[${this.pluginName}] Coordinating file: ${fileId}`);
+    try {
+      this.publish('checking/file', {
+        fileId: fileId
```
Comment (VolodymyrFihurniak) [inline plugins/FileCoordinatorPlugin/src/index.ts:20]: The key in the object should be: `file_id`

File: services/file-coordinator/src/utils/connectionValidator.ts
Comment (VolodymyrFihurniak) [inline services/file-coordinator/src/utils/connectionValidator.ts:1]: I think it's still worth checking the server connection

Comment (denbykov) [reply services/file-coordinator/src/utils/connectionValidator.ts:1]: I do not care about it. If you want it, you can implement it

**Review by denbykov — APPROVED** [review]

**Pull request merged**